File: computepatch_src/PatchAtom.py
from abc import ABCMeta, abstractmethod

# Classe abstraite Parent

class PatchAtom:
    def __init__(self, line_nb):
        self.line_number = line_nb

    def __eq__(self, other):
        return type(other) is type(self) and self.__dict__ == other.__dict__

    def __repr__(self):
        return '%s(%d)' % (self.__class__.__name__, self.line_number)

# Classes héritant de PatchAtom

class AdditionAtom(PatchAtom):
    def __init__(self, line_nb, new_line):
        super().__init__(line_nb)
        assert isinstance(new_line, str)
        self.new_line = new_line
        self.cost = 10 + len(self.new_line)

    def __str__(self):
        return "+ {}\n{}".format(self.line_number, self.new_line.replace('\n', ''))

    @staticmethod
    def compute_cost(line):
        return 10 + len(line)

class SubstituteAtom(PatchAtom):
    def __init__(self, line_nb, subs_line):
        super().__init__(line_nb)
        assert isinstance(subs_line, str)
        self.subs_line = subs_line
        self.cost = 10 + len(self.subs_line)

    def __str__(self):
        return "= {}\n{}".format(self.line_number, self.subs_line.replace('\n', ''))

    @staticmethod
    def compute_cost(line):
        return 10 + len(line)

class DestructionAtom(PatchAtom):
    cost = 10
    def __init__(self, line_nb):
        super().__init__(line_nb)

    def __str__(self):
        return "d {}".format(self.line_number)

    @staticmethod
    def compute_cost():
        return 10

class DestructionMultAtom(PatchAtom):
    cost = 15
    def __init__(self, line_nb, destruction_nb):
        super().__init__(line_nb)
        self.destruction_nb = destruction_nb

    def __str__(self):
        return "D {} {}".format(self.line_number, self.destruction_nb)

    @staticmethod
    def compute_cost():
        return 15

File: computepatch_c++/patchAtom.cpp
#include "patchAtom.hpp"

PatchAtom::PatchAtom(int line_nb) {
    this->line_nb = line_nb ;
}

IdentityAtom::IdentityAtom(int line_nb) : PatchAtom(line_nb) {

}

int IdentityAtom::compute_cost(void) const {
    return 0 ;
}

void IdentityAtom::printInStream(std::ostream &stream) const {

}

AdditionAtom::AdditionAtom(int line_nb, const std::string &new_line) : PatchAtom(line_nb) {
    this->new_line = new_line ;
}

int AdditionAtom::compute_cost(void) const {
    return 10 + new_line.size() ;
}

void AdditionAtom::printInStream(std::ostream &stream) const {
    stream << "+" << this->line_nb << "\n" << this->new_line << "\n" ;
}

SubstituteAtom::SubstituteAtom(int line_nb, const std::string &subs_line) : PatchAtom(line_nb) {
    this->subs_line = subs_line ;
}

int SubstituteAtom::compute_cost(void) const {
    return 10 + subs_line.size() ;
}

void SubstituteAtom::printInStream(std::ostream &stream) const {
    stream << "=" << this->line_nb << "\n" << this->subs_line << "\n" ;
}

DestructionAtom::DestructionAtom(int line_nb) : PatchAtom(line_nb) {

}

int DestructionAtom::compute_cost(void) const {
    return 10 ;
}

void DestructionAtom::printInStream(std::ostream &stream) const {
    stream << "d" << this->line_nb << "\n" ;
}

DestructionMultAtom::DestructionMultAtom(int line_nb, int size) : PatchAtom(line_nb) {
    this->size = size ;
}

int DestructionMultAtom::compute_cost(void) const {
    return 15 ;
}

void DestructionMultAtom::printInStream(std::ostream &stream) const {
    stream << "D" << this->line_nb << " " << this->size << "\n" ;
}
